ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,851 @@
+   * @param waitTimeSecs the number of seconds to wait for shutdown to complete.
+   * @throws InterruptedException if the wait for shutdown was interrupted.
+   */
+  void shutdown(long waitTimeSecs) throws InterruptedException {
+    isActive = false;
+    if (waitTimeSecs > 0 && !runningLatch.await(waitTimeSecs, TimeUnit.SECONDS)) {
+      logger.error("Compactor did not shutdown within 2 seconds");
+    }
+  }
+
+  /**
+   * Compacts the store by copying valid data from the log segments in {@code details} to swap spaces and switching
+   * the compacted segments out for the swap spaces. Returns once the compaction is complete.
+   * @param details the {@link CompactionDetails} for the compaction.
+   * @throws IllegalArgumentException if any of the provided segments doesn't exist in the log, if one or more offsets
+   * in the segments to compact are in the journal or if the last modified time of any of the provided log segments is
+   * greater than the provided reference time.
+   * @throws IllegalStateException if the compactor has not been started.
+   * @throws InterruptedException if the compaction was interrupted
+   * @throws IOException if there were I/O errors during copying, commit or cleanup.
+   * @throws StoreException if there were exceptions reading to writing to store components.
+   */
+  void compact(CompactionDetails details) throws InterruptedException, IOException, StoreException {
+    if (srcIndex == null) {
+      throw new IllegalStateException("Compactor has not been started");
+    }
+    checkSanity(details);
+    compactionLog = new CompactionLog(dataDir.getAbsolutePath(), storeId, time, details);
+    resumeCompaction();
+  }
+
+  /**
+   * Resumes compaction from where it was left off.
+   * @throws IllegalStateException if the compactor has not been started or if there is no compaction to resume.
+   * @throws InterruptedException if the compaction was interrupted
+   * @throws IOException if there were I/O errors during copying, commit or cleanup.
+   * @throws StoreException if there were exceptions reading to writing to store components.
+   */
+  void resumeCompaction() throws InterruptedException, IOException, StoreException {
+    if (srcIndex == null) {
+      throw new IllegalStateException("Compactor has not been started");
+    } else if (compactionLog == null) {
+      throw new IllegalStateException("There is no compaction to resume");
+    }
+    runningLatch = new CountDownLatch(1);
+    while (isActive && !compactionLog.getCompactionPhase().equals(CompactionLog.Phase.DONE)) {
+      CompactionLog.Phase phase = compactionLog.getCompactionPhase();
+      switch (phase) {
+        case PREPARE:
+          compactionLog.markCopyStart();
+        case COPY:
+          copy();
+          if (isActive) {
+            compactionLog.markCommitStart();
+            commit(false);
+            compactionLog.markCleanupStart();
+            waitTillRefCountZero();
```

Comment:
I think the `isActive` checks are in the right places but let me document my intentions so that they can be verified.
1. Wait for ref count to become 0 before cleanup if `isActive` is `true` (so that requests don't get dropped), otherwise don't because shutting down quickly is the main goal (requests in progress will be dropped anyway).
2. Stop at a safe point during copying (in between copying records and not in the middle of one).
3. Stop inbetween compaction cycles if `isActive` is `false`.
4. Stop before commit starts if `isActive` is `false`. Once commit starts, do cleanup  and finish the current cycle even if the value of `isActive` changes in the middle (the operations are fast).